test/jdk/sun/security/pkcs11/PKCS11Test.java

```diff
@@ -222,61 +225,68 @@ public static void main(PKCS11Test test, String[] args) throws Exception {
             // useful for ./Provider/Login.sh, where a SecurityManager exists.
             if (oldProviders.length == newProviders.length) {
                 found = false;
-                for (int i = 0; i<oldProviders.length; i++) {
+                for (int i = 0; i < oldProviders.length; i++) {
                     if (oldProviders[i] != newProviders[i]) {
                         found = true;
                         break;
                     }
                 }
             }
             if (found) {
-                for (Provider p: newProviders) {
+                for (Provider p : newProviders) {
                     Security.removeProvider(p.getName());
                 }
-                for (Provider p: oldProviders) {
+                for (Provider p : oldProviders) {
                     Security.addProvider(p);
                 }
             }
         }
     }
 
     public static void testDeimos(PKCS11Test test) throws Exception {
-        if (new File("/opt/SUNWconn/lib/libpkcs11.so").isFile() == false ||
-            "true".equals(System.getProperty("NO_DEIMOS"))) {
+        System.out.println("===> testDeimos: Starting test run");
+        if ("true".equals(System.getProperty("NO_DEIMOS"))) {
+            System.out.println("Skip Deimos software as test configured with NO_DEIMOS");
             return;
         }
+
+        if (!new File("/opt/SUNWconn/lib/libpkcs11.so").isFile()) {
+            throw new SkippedException("testDeimos: \"/opt/SUNWconn/lib/libpkcs11.so\" " +
+                    "file required for Deimos not found");
+        }
+
         String base = getBase();
         String p11config = base + SEP + "nss" + SEP + "p11-deimos.txt";
         Provider p = getSunPKCS11(p11config);
         test.premain(p);
+        System.out.println("testDeimos: Completed");
     }
 
+    // Run test for default configured PKCS11 providers (if any)
     public static void testDefault(PKCS11Test test) throws Exception {
-        // run test for default configured PKCS11 providers (if any)
+        System.out.println("===> testDefault: Starting test run");
+        boolean foundPKCS11 = false;
 
         if ("true".equals(System.getProperty("NO_DEFAULT"))) {
+            System.out.println("Skip default provider as test configured with NO_DEFAULT");
             return;
         }
 
         Provider[] providers = Security.getProviders();
-        for (int i = 0; i < providers.length; i++) {
-            Provider p = providers[i];
+        for (Provider p : providers) {
             if (p.getName().startsWith("SunPKCS11-")) {
```

Comment:
Please correct me, if i am wrong. But as per my understanding there can be many PKCS11 provider instance exist in same time based on different token configuration and the order can be different too. In that case specifying the default PKCS11 provider could be better which is 'p.getName().equals("SunPKCS11")'. Here it is safe because testDefault() is the 1st method to execute. Also the loop should break immediately after finding the default provider.

Reply:
As is currently, test is run for all found providers. I am avoiding to update existing test logic with this change. I will consider this update while working on [JDK-8313575](https://bugs.openjdk.org/browse/JDK-8313575). More TBD